lib/Conversion/CalyxToFSM/CalyxToFSM.cpp

```diff
@@ -268,6 +268,71 @@ LogicalResult CompileFSMVisitor::visit(StateOp currentState, EnableOp enableOp,
   return success();
 }
 
+// CompileInvoke is used to convert invoke operations to group operations and
+// enable operations.
+class CompileInvoke {
+public:
+  CompileInvoke(ComponentOp component, OpBuilder builder)
+      : component(component), builder(builder) {}
+  void compile();
+
+private:
+  void lowerInvokeOp(InvokeOp invokeOp);
+  ComponentOp component;
+  OpBuilder builder;
+  // It is used to pass to the go port and assign a value to the go port.
+  hw::ConstantOp constantOp = nullptr;
+  llvm::StringMap<size_t> groupNameMap;
+};
+
+// Access all invokeOp.
+void CompileInvoke::compile() {
+  llvm::SmallVector<InvokeOp> invokeOps =
+      component.getControlOp().getInvokeOps();
+  for (InvokeOp op : invokeOps) {
+    lowerInvokeOp(op);
+  }
+}
+
+// Convert an invoke operation to a group operation and an enable operation.
+void CompileInvoke::lowerInvokeOp(InvokeOp invokeOp) {
+  // Create a ConstantOp to assign a value to the go port.
+  if (!constantOp) {
```

Comment:
I actually did eliminate this optimization by creating a ConstantOp directly, without judging the previous ConstantOp. Since there may be multiple invokes in a component, and I created the ConstantOp when I processed the first invoke, what I'm judging here is whether or not the ConstantOp is empty, and if it's empty, that means I'm processing the first invoke in the component, and subsequent invokes don't need to have a ConstantOp created(If the user uses invoke, a ConstantOp is created in a component.).

Reply:
What I'm suggesting is get rid of this extra `constantOp` state and just create a new `ConstantOp` each time you need one with the builder.

Reply:
Honestly, in that case a constantOp is created for each invoke, and in the end the code generated by lower-calyx-to-fsm doesn't look very elegant (maybe I still don't understand what you mean?).

Reply:
I think further discussion is needed here.

Reply:
> in the end the code generated by lower-calyx-to-fsm doesn't look very elegant

I think the point here is this is IR, not code. Constant ops are extremely common, and are trivially handled by CSE, which most compiler pipelines run at some point. I agree with Chris that it's worth trading simplicity of implementation for some extra ops in the IR--they will disappear after CSE.